src/gneiss/server/cbe/component.adb

```diff
@@ -0,0 +1,824 @@
+            return;
+         end if;
+         Block_Client.Initialize (Client, Cap, "CBE", 42);
+         if not Block.Initialized (Client) then
+            Gns.Log.Client.Error (Log, "Client initialization failed.");
+            Main.Vacate (Cap, Main.Failure);
+            return;
+         end if;
+         Timer_Client.Initialize (Timer, Cap);
+         if not Gns.Timer.Initialized (Timer) then
+            Gns.Log.Client.Error (Log, "Timer initialization failed");
+            Main.Vacate (Cap, Main.Failure);
+            return;
+         end if;
+         Read_Superblock (Progress);
+         Initialize_Crypto;
+         Timer_Client.Set_Timeout (Timer, 1.0);
+         --  snapshot every 1.0s
+      else
+         Main.Vacate (Cap, Main.Failure);
+      end if;
+   end Construct;
+
+   procedure Destruct is
+   begin
+      if Gns.Log.Initialized (Log) then
+         Gns.Log.Client.Warning (Log, "Stopping CBE");
+      end if;
+      Gns.Log.Client.Finalize (Log);
+      Block_Client.Finalize (Client);
+      Block_Dispatcher.Finalize (Dispatcher);
+      Timer_Client.Finalize (Timer);
+   end Destruct;
+
+   procedure Initialize_Crypto
+   is
+      Success : Boolean;
+   begin
+      External.Crypto.Set_Key
+         (Crypto, 1, 0, External.Crypto.Key_Data_Type'(others => 42), Success);
+      if Success then
+         External.Crypto.Set_Key
+            (Crypto, 2, 0, External.Crypto.Key_Data_Type'(others => 42),
+             Success);
+      end if;
+      if not Success then
+         Gns.Log.Client.Error (Log, "Failed to set crypto key");
+         Main.Vacate (Capability, Main.Failure);
+      end if;
+   end Initialize_Crypto;
+
+   procedure Write
+      (C : in out Block.Client_Session;
+       I :        Request_Id;
+       D :    out Buffer)
+   is
+      pragma Unreferenced (C);
+   begin
+      Convert_Block (Io_Buffer (C_Cache (I).I), D);
+   end Write;
+
+   procedure Read
+      (C : in out Block.Client_Session;
+       I :        Request_Id;
+       D :        Buffer)
+   is
+      pragma Unreferenced (C);
+      use type Block.Request_Status;
+      S : CBE.Superblocks_Index_Type;
+   begin
+      if Ready then
+         Convert_Block (D, Io_Buffer (C_Cache (I).I));
+      else --  reading superblocks
+         if D'Length = 4096 and then Block_Client.Status (Sbsr (I)) = Block.Ok
+            and then Block_Client.Start (Sbsr (I)) in 0 .. 7
+         then
+            S := CBE.Superblocks_Index_Type (Block_Client.Start (Sbsr (I)));
+            Convert_Block (D, Superblocks (S));
+            Superblocks_Init (S) := True;
+         end if;
+      end if;
+   end Read;
+
+   procedure Snapshot_Event
+   is
+      function Image is new Gns.Strings_Generic.Image_Modular
+         (CBE.Generation_Type);
+      Snapshot : CBE.Generation_Type;
+      Success  : Boolean;
+   begin
+      CBE.Library.Create_Snapshot (Cbe_Session, False, Snapshot, Success);
+      if not Success then
```

Comment:
Sidenote: `CBE.Library.Create_Snapshot` will also return `False` in case there is no need to create a snapshot, i.e. there are no changes to the CBE and you have to poll `CBE.Library.Snapshot_Creation_Complete` because it might take some time for the snapshot to be created (all outstanding changes must be written to disk).

Reply:
Fixed by https://github.com/cbe-project/cbe/pull/39/commits/5b04cf7d0dbc857b697ee864370b318014376000.